File: Leetcode_POTD_2025/8_August/808_Soup_Servings.py
class Solution(object):
    def soupServings(self, n):
        if n >= 4800:
            return 1.0
        m = (n + 24) // 25
        memo = {}
        return self._calc_prob(m, m, memo)

    def _calc_prob(self, a, b, memo):
        if (a, b) in memo:
            return memo[(a, b)]
        if a <= 0 and b <= 0:
            return 0.5
        if a <= 0:
            return 1.0
        if b <= 0:
            return 0.0

        res = 0.25 * (self._calc_prob(a - 4, b, memo) +
                      self._calc_prob(a - 3, b - 1, memo) +
                      self._calc_prob(a - 2, b - 2, memo) +
                      self._calc_prob(a - 1, b - 3, memo))
        memo[(a, b)] = res
        return res

File: Leetcode_POTD_2025/8_August/808_Soup_Servings.java
class Solution {
    private double[][] memo;
    public double soupServings(int n) {
        if (n >= 4800) return 1.0;
        int size = (n + 24) / 25;
        memo = new double[size + 1][size + 1];
        return calcProb(size, size);
    }
    private double calcProb(int a, int b) {
        if (a <= 0 && b <= 0) return 0.5;
        if (a <= 0) return 1.0;
        if (b <= 0) return 0.0;
        if (memo[a][b] > 0) return memo[a][b];
        double res = 0.25 * (calcProb(a - 4, b) +
                             calcProb(a - 3, b - 1) +
                             calcProb(a - 2, b - 2) +
                             calcProb(a - 1, b - 3));
        return memo[a][b] = res;
    }
}
